difftest/offline_difftest/src/difftest.rs

```diff
@@ -0,0 +1,126 @@
+mod dut;
+mod spike;
+
+use dut::*;
+use serde::Deserialize;
+pub use spike::SpikeHandle;
+use std::path::Path;
+use tracing::error;
+
+pub struct Difftest {
+	spike: SpikeHandle,
+	dut: Dut,
+	config: Config,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct LsuBankParameter {
+	name: String,
+	region: String,
+	beatbyte: u32,
+	accessScalar: bool,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct LogicModuleParametersParameter {
+	datapathWidth: u32,
+	latency: u32,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct LogicModuleParameter {
+	parameter: LogicModuleParametersParameter,
+	generator: String,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct OtherModuleParametersParameter {
+	datapathWidth: u32,
+	vlMaxBits: u32,
+	groupNumberBits: u32,
+	laneNumberBits: u32,
+	dataPathByteWidth: u32,
+	latency: u32,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct OtherModuleParameters {
+	parameter: OtherModuleParametersParameter,
+	generator: String,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct VfuInstantiateParameter {
+	slotCount: u32,
+	logicModuleParameters: Vec<(LogicModuleParameter, Vec<u32>)>,
+	aluModuleParameters: Vec<(LogicModuleParameter, Vec<u32>)>,
+	shifterModuleParameters: Vec<(LogicModuleParameter, Vec<u32>)>,
+	mulModuleParameters: Vec<(LogicModuleParameter, Vec<u32>)>,
+	divModuleParameters: Vec<(LogicModuleParameter, Vec<u32>)>,
+	divfpModuleParameters: Vec<(LogicModuleParameter, Vec<u32>)>,
+	floatModuleParameters: Vec<(LogicModuleParameter, Vec<u32>)>,
+	otherModuleParameters: Vec<(OtherModuleParameters, Vec<u32>)>,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct Parameter {
```

Comment:
In the future it won’t be read from the json config. But the refactor won’t be large.

Reply:
Those metadatas will use omreader reading from the mlirbc